Frameworks/Foundation/NSOperationQueue.mm

```diff
@@ -30,167 +30,320 @@
+                    // one and it could finish muc later (not synchronously executed with [operation start]).
+                    _curOperation = nil;
+                }
+            } else {
+                {
+                    std::lock_guard<std::mutex> lock(_lock);
+                    // Not yet ready. Demote it to the holding area. Will pick it up again later.
+                    _operationsHoldingArea[priority].emplace_back(std::move(candidateOperation));
+                }
+
+                // We claim that we have done some work here. This keeps the queue pumping and
+                // will eventually bring this item back out for consideration and will continue
+                // through this list for anyother ready operations.
+                return TRUE;
+            }
+        }
 
-    NSOperationQueuePriv() {
-        memset(this, 0, sizeof(NSOperationQueuePriv));
-        _threadRunningLock = PTHREAD_MUTEX_INITIALIZER;
+        return candidateOperation != nil;
     }
-};
 
-extern pthread_key_t g_currentDispatchQueue;
+    void WaitUntilNotSuspended() {
+        std::unique_lock<std::mutex> lock(_lock);
+        while (_isSuspended) {
+            _suspendedCondition.wait(lock);
+        }
+        lock.unlock();
+    }
 
-static inline gpointer CompareExchangePointer(volatile gpointer* dest, gpointer exch, gpointer comp) {
-    return (gpointer)EbrCompareExchange((volatile int*)dest, (long)exch, (long)comp);
-}
+    void ClearCandidateLists() {
+        std::lock_guard<std::mutex> lock(_lock);
+        for (int i = 0; i < NSOperationQueuePriority_Count; i++) {
+            _candidateOperationsToRunNext[i].clear();
+        }
+    }
 
-NSAtomicListRef NSAtomicListSteal(NSAtomicListRef* listPtr) {
-    NSAtomicListRef ret;
-    do {
-        ret = *listPtr;
-    } while (CompareExchangePointer((void**)listPtr, NULL, ret) != ret);
-    return ret;
-}
+    BOOL HasMoreWork() {
+        std::lock_guard<std::mutex> lock(_lock);
 
-void NSAtomicListReverse(NSAtomicListRef* listPtr) {
-    struct NSAtomicListNode* cur = *listPtr;
-    struct NSAtomicListNode* prev = NULL;
-    struct NSAtomicListNode* next = NULL;
+        BOOL toReturn = _HasMoreWorkNoLock();
+
+        if (!toReturn) {
+            // No work left to do. Clear out the thread here. This signals that a new work thread is needed to be created and
+            // current one can end its execution.
+            _thread = nil;
+        }
 
-    if (!cur) {
-        return;
+        return toReturn;
     }
 
-    do {
-        next = cur->next;
-        cur->next = prev;
+    void AddOperation(NSOperation* operation, unsigned int priority, NSOperationQueue* queue, SEL selector) {
+        {
+            std::lock_guard<std::mutex> lock(_lock);
 
-        if (next) {
-            prev = cur;
-            cur = next;
+            StrongId<NSOperation> strongOperation = operation;
+            _operationsHoldingArea[priority].emplace_back(strongOperation);
```

Comment:
lol oops. I meant to std::move. It doesn't really matter though too much.

---
In reply to: [95450127](https://github.com/Microsoft/WinObjC/pull/1676#discussion_r95450127) [](ancestors = 95450127)